EventVisualisation/Workflow/src/EveWorkflowHelper.cxx

```diff
@@ -13,167 +13,205 @@
-        o2::its::ioutils::convertCompactClusters(clusITS, pattIt, retVal->ITSClustersArray, *calib->itsPatternDict);
-      }
+  auto creator = [maskTrk, this](auto& trk, GID gid, float time, float) {
+    if (!maskTrk[gid.getSource()]) {
+      return true;
     }
+    mTrackSet.trackGID.push_back(gid);
+    mTrackSet.trackTime.push_back(time);
+    return true;
+  };
+  this->mRecoCont.createTracksVariadic(creator);
+}
+
+void EveWorkflowHelper::draw(std::string jsonPath, int numberOfFiles, int numberOfTracks)
+{
+  EveWorkflowHelper::prepareITSClusters();
+
+  size_t nTracks = mTrackSet.trackGID.size();
+  if (numberOfTracks != -1 && numberOfTracks < nTracks) {
+    nTracks = numberOfTracks; // less than available
   }
+  for (size_t it = 0; it < nTracks; it++) {
+    const auto& gid = mTrackSet.trackGID[it];
+    auto tim = mTrackSet.trackTime[it];
 
-  if (maskTrk[GID::ITSTPC] && nTracksTPCITSO2 == 0) {
-    const auto& trkITSTPC = recoCont.getTPCITSTracks();
-    if (trkITSTPC.size()) {
-      nTracksTPCITSO2 = trkITSTPC.size();
+    if (gid.getSource() == GID::TPC) {
+      drawTPC(gid, tim);
+    }
+    if (gid.getSource() == GID::ITS) {
+      drawITS(gid, tim);
+    }
+    if (gid.getSource() == GID::ITSTPC) {
+      drawITSTPC(gid, tim);
+    } else if (gid.getSource() == GID::ITSTPCTOF) {
+      drawITSTPCTOF(gid, tim);
     }
-    LOG(info) << "Got " << nTracksTPCITSO2 << " ITS-TPC Tracks";
   }
+  FileProducer producer(jsonPath, numberOfFiles);
+  mEvent.toFile(producer.newFileName());
+}
 
-  if ((maskMatch[GID::TOF] || maskMatch[GID::ITSTPCTOF])) {
-    const auto& tofMatches = recoCont.getITSTPCTOFMatches();
-    if (tofMatches.size()) {
-      nITSTPCTOFMatches = tofMatches.size();
-    }
-    LOG(info) << "Got " << nITSTPCTOFMatches << " ITS-TPC-TOF Matches";
+void EveWorkflowHelper::drawITSTPC(GID gid, float trackTime)
+{
+  const auto& track = mRecoCont.getTPCITSTrack(gid);
+  auto pnts = getTrackPoints(track, minmaxR[gid.getSource()].first, minmaxR[gid.getSource()].second, 4);
+  addTrackToEvent([this, trackTime](GID gid) { return mRecoCont.getTPCITSTrack(gid); }, trackTime, 0.);
+  GID gidTPC = track.getRefTPC();
+  GID gidITS = track.getRefITS();
+  drawITSClusters(gidITS, trackTime);
+  drawTPCClusters(gidTPC, trackTime);
+}
+
+void EveWorkflowHelper::drawITSTPCTOF(GID gid, float trackTime)
+{
+  const auto& track = mRecoCont.getITSTPCTOFTrack(gid);
+  addTrackToEvent([this, trackTime](GID gid) { return mRecoCont.getITSTPCTOFTrack(gid); }, trackTime, 0.);
+  GID gidTPC = track.getRefTPC();
+  GID gidITS = track.getRefITS();
+
+  drawITSClusters(gidITS, trackTime);
+  drawTPCClusters(gidTPC, trackTime);
+}
+
+std::vector<PNT> EveWorkflowHelper::getTrackPoints(const o2::track::TrackPar& trc, float minR, float maxR, float maxStep)
+{
+  // adjust minR according to real track start fro track starting point
+  float rMin = std::sqrt(trc.getX() * trc.getX() + trc.getY() * trc.getY());
+  if (rMin > minR)
+    minR = rMin;
```

Comment:
out codechecker requires  `{}` on different lines even for potential oneliners

Reply:
done